CNV-27612: Procedure to build VM container disk image (#59491)
openshift/openshift-docs: modules/virt-building-vm-containerdisk-image.adoc

// Module included in the following assemblies:
//
// * virt//support/monitoring/virt-running-cluster-checkups.adoc              

:_content-type: PROCEDURE                                             
[id="virt-building-vm-containerdisk-image_{context}"]                                  
= Building a container disk image for RHEL virtual machines

> You can build a custom RHEL 8 OS image in `qcow2` format and use it to build a container disk image for RHEL virtual machines (VMs). You can store the container disk image in a registry that is accessible from your cluster and specify the image location in the `spec.param.vmContainerDiskImage` attribute of the DPDK checkup config map. 

Reviewer: ```suggestion
You can build a custom RHEL 8 OS image in `qcow2` format and use it to create a container disk image. You can store the container disk image in a registry that is accessible from your cluster and specify the image location in the `spec.param.vmContainerDiskImage` attribute of the DPDK checkup config map.
```

> Run the following procedure in the image builder VM to build a container disk image for RHEL 8.7 VMs with 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory. 

Reviewer: I suggest to define the term "image builder VM" as a RHEL 8 VM used to build custom RHEL images.

.Prerequisites
* You have access to the image builder tool and its CLI (`composer-cli`).
* You have verified that you can build a RHEL 8.7 image:
+
[source,terminal]
----
# composer-cli distros list
----
+
[NOTE]
====
To run the `composer-cli` commands as non-root, add your user to the `weldr` or `root` groups:

[source,terminal]
----
# usermod -a -G weldr user
----
[source,terminal]
----
$ newgrp weldr
----
====

> * You have installed the `virt-customize` tool.

Reviewer: Please consider adding a link that explains how to install the `virt-customize` tool.
Alternatively, the command to install it on RHEL 8 is:
```bash
# dnf install libguestfs-tools
```

+
[source,terminal]
----
# composer-cli compose status
----

. Download the `qcow2` image file by specifying its UUID:
+
[source,terminal]
----
# composer-cli compose image <UUID>
----

. Create the customization scripts:
+
[source,terminal]
----
$ cat <<EOF >customize-vm
echo  isolated_cores=2-7 > /etc/tuned/cpu-partitioning-variables.conf
tuned-adm profile cpu-partitioning
echo "options vfio enable_unsafe_noiommu_mode=1" > /etc/modprobe.d/vfio-noiommu.conf
EOF
----
+
[source,terminal]
----
$ cat <<EOF >first-boot
driverctl set-override 0000:06:00.0 vfio-pci
driverctl set-override 0000:07:00.0 vfio-pci

mkdir /mnt/huge
mount /mnt/huge --source nodev -t hugetlbfs -o pagesize=1GB
EOF
----

. Customize the base OS image by using the `virt-customize` tool:
+
----
$ virt-customize -a <UUID>.qcow2 --run=customize-vm --firstboot=first-boot --selinux-relabel
----

. Create a Dockerfile that contains all the commands to build the container disk image:
+
[source,terminal]
----
$ cat << EOF > Dockerfile
FROM scratch
COPY <uuid>-disk.qcow2 /disk/
EOF
----

. Build and tag the container:
+
[source,terminal]
----

> $ sudo podman build . -t dpdk-rhel:latest

Reviewer: There is no need for using `sudo`.
It was a mistake in the upstream documentation.
```suggestion
$ podman build . -t dpdk-rhel:latest
```

----

. Push the container disk image to a registry that is accessible to your cluster:
+
[source,terminal]
----

> $ sudo podman push dpdk-rhel:latest

Reviewer: There is no need for using sudo.
It was a mistake in the upstream documentation.

```suggestion
$ podman push dpdk-rhel:latest
```

// Module included in the following assemblies:
//
// * virt//support/monitoring/virt-running-cluster-checkups.adoc              

:_content-type: PROCEDURE                                             
[id="virt-building-vm-containerdisk-image_{context}"]                                  
= Building a container disk image for RHEL virtual machines

You can build a custom RHEL 8 OS image in `qcow2` format and use it to build a container disk image for RHEL virtual machines (VMs). You can store the container disk image in a registry that is accessible from your cluster and specify the image location in the `spec.param.vmContainerDiskImage` attribute of the DPDK checkup config map. 

> Use the following procedure to build a container disk image for RHEL 8.7 VMs with 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory.

Author: @orelmisan There is existing RHEL documentation that lists the steps to [install the image builder tool](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/composing_a_customized_rhel_system_image/installing-composer_composing-a-customized-rhel-system-image), including the prerequisties that you have listed. I did not want to duplicate those steps, so I added a link to the documentation in the Additional resources section.

I have also linked to the [Knowledgebase article](https://access.redhat.com/solutions/253273) about using Red Hat Subscription Manager.

> Use the following procedure to build a container disk image for RHEL 8.7 VMs with 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory.

Author: I will change the wording to say that the procedure is executed in the builder VM

> Use the following procedure to build a container disk image for RHEL 8.7 VMs with 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory.

Reviewer: Thank you @sjhala-ccs.

.Prerequisites
* You have access to the image builder tool and its CLI (`composer-cli`).
* You have installed the `virt-customize` tool.
* You have installed `podman`.

> * You have verified that you can use the RHEL 8.7 distro:

Reviewer: ```suggestion
* You have verified that you can build a RHEL 8.7 image:
```

+
[source,terminal]
----
# composer-cli distros list
----

.Procedure
. Create an image blueprint file in TOML format which contains the packages to be installed, kernel customizations, and the services to be disabled during boot time:
+
[source,terminal]
----
$ cat << EOF > dpdk-vm.toml
name = "dpdk_image"
description = "Image to use with the DPDK checkup"
version = "0.0.1"
distro = "rhel-87"

[[packages]]
name = "dpdk"

[[packages]]
name = "dpdk-tools"

[[packages]]
name = "driverctl"

[[packages]]
name = "tuned-profiles-cpu-partitioning"

[customizations.kernel]
append = "default_hugepagesz=1GB hugepagesz=1G hugepages=8 isolcpus=2-7"

[customizations.services]
disabled = ["NetworkManager-wait-online", "sshd"]
EOF
----

. Push the blueprint file to the image builder tool by using its CLI:
+
[source,terminal]
----
# composer-cli blueprints push dpdk-vm.toml
----
+
[NOTE]
====
To run the `composer-cli` commands as non-root, add your user to the `weldr` or `root` groups:

[source,terminal]
----
# usermod -a -G weldr user
----
[source,terminal]
----
$ newgrp weldr
----

> ====

Reviewer: Please consider moving this note before the first usage of `composer-cli`.

// Module included in the following assemblies:
//
// * virt//support/monitoring/virt-running-cluster-checkups.adoc              

:_content-type: PROCEDURE                                             
[id="virt-building-vm-containerdisk-image_{context}"]                                  
= Building a container disk image for RHEL virtual machines

You can build a custom RHEL 8 OS image in `qcow2` format and use it to create a container disk image. You can store the container disk image in a registry that is accessible from your cluster and specify the image location in the `spec.param.vmContainerDiskImage` attribute of the DPDK checkup config map. 

> Run the following procedure in the image builder virtual machine (VM) terminal to build a container disk image for RHEL 8.7 VMs with 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory. The _image builder VM_ is a RHEL 8 VM that can be used to build custom RHEL images.

Reviewer: ```suggestion
In order to perform this procedure, you will have to create an image builder virtual machine.
The _image builder VM_ is a RHEL 8 VM that can be used to build custom RHEL images.
The image builder VM should run RHEL 8.7 and should have at least the following spec:
 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory.
```

Reviewer 2: Taking @orelmisan 's revision into consideration, I would suggest these slight revisions: 

In order to perform this procedure, you must create an image builder virtual machine (VM). The image builder VM is a {op-system-base} 8 VM that can be used to build custom {op-system-base} images.

I would move 

The image builder VM should run RHEL 8.7 and should have at least the following spec:
 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory

to prerequisites

.Prerequisites
* Install the image builder tool and its CLI (`composer-cli`).

> * Verify that you can build a RHEL 8.7 image:

Reviewer 2: .Prerequisites 

* You have installed the image builder tool and its CLI (`composer-cli`). 
* Your image builder VM is running {op-system-base} 8.7.
* Your system has a minimum of 2 CPU cores.
* Your system has a minimum of 4 GiB of ram.
* Your system has a minimum of 20 GB of storage in the `var` directory.
* You have installed the `virt-customize` tool. 
* You have installed the Podman CLI tool (`podman`).

+

> [source,terminal]

Reviewer 2: If warranted, I would add this as Step 1 in the procedure. We shouldn't have a one-step procedure in the prerequisites. This could be either Step 1, or an Optional Step, in the procedure portion of the document.

Author: Moved it to the procedure as step 1

----
# composer-cli distros list
----
+
[NOTE]
====
To run the `composer-cli` commands as non-root, add your user to the `weldr` or `root` groups:

[source,terminal]
----
# usermod -a -G weldr user
----
[source,terminal]
----
$ newgrp weldr
----
====
* Install the `virt-customize` tool by using the following command:
+
[source,terminal]
----
# dnf install libguestfs-tools
----
* Install `podman`.

> .Procedure

Reviewer 2: Not necessary, but you might want to put a space between .Procedure and the first step.

> . Create an image blueprint file in TOML format which contains the packages to be installed, kernel customizations, and the services to be disabled during boot time:

Reviewer 2: ```suggestion
. Enter the following command to create an image blueprint file in TOML format that contains the packages to be installed, kernel customizations, and the services to be disabled during boot time:
```

+
[source,terminal]
----
$ cat << EOF > dpdk-vm.toml
name = "dpdk_image"
description = "Image to use with the DPDK checkup"
version = "0.0.1"
distro = "rhel-87"

[[packages]]
name = "dpdk"

[[packages]]
name = "dpdk-tools"

[[packages]]
name = "driverctl"

[[packages]]
name = "tuned-profiles-cpu-partitioning"

[customizations.kernel]
append = "default_hugepagesz=1GB hugepagesz=1G hugepages=8 isolcpus=2-7"

[customizations.services]
disabled = ["NetworkManager-wait-online", "sshd"]
EOF
----

> . Push the blueprint file to the image builder tool by using its CLI:

Reviewer 2: ```suggestion
. Enter the following command to push the blueprint file to the image builder tool:
```

+
[source,terminal]
----
# composer-cli blueprints push dpdk-vm.toml
----

> . Generate the system image by specifying the blueprint name and output file format. The Universally Unique Identifier (UUID) of the image is displayed when you start the compose process.

Reviewer 2: ```suggestion
. Enter the following command to generate the system image by specifying the blueprint name and output file format. The Universally Unique Identifier (UUID) of the image is displayed when you start the compose process.
```

If you're going to mention that the UUID of the image is displayed, you might want to provide an ".Example output". Just a thought.

Author: @orelmisan Do you think it would be helpful to provide an example here to show the UUID of an image?

Reviewer: Its just a long number of letters and digits.
I don't think it matters.

+
[source,terminal]
----
# composer-cli compose start dpdk_image qcow2
----

> . Wait for the compose process to complete. The compose status must show `FINISHED`.

Reviewer 2: ```suggestion
. Enter the following command to check the compose status. Before continuing to the next step, the compose status must show `FINISHED`. 
```

+
[source,terminal]
----
# composer-cli compose status
----

> . Download the `qcow2` image file by specifying its UUID:

Reviewer 2: ```suggestion
. Enter the following command to download the `qcow2` image file by specifying its UUID:
```

+
[source,terminal]
----
# composer-cli compose image <UUID>
----

> . Create the customization scripts:

Reviewer 2: ```suggestion
. Create the following customization scripts by entering the following command:
```

^ With this suggestion, I'm hoping that it'll be more clear to readers that there are x2 customization scripts they must enter. Might not be warranted.

+
[source,terminal]
----
$ cat <<EOF >customize-vm
echo  isolated_cores=2-7 > /etc/tuned/cpu-partitioning-variables.conf
tuned-adm profile cpu-partitioning
echo "options vfio enable_unsafe_noiommu_mode=1" > /etc/modprobe.d/vfio-noiommu.conf
EOF
----
+
[source,terminal]
----
$ cat <<EOF >first-boot
driverctl set-override 0000:06:00.0 vfio-pci
driverctl set-override 0000:07:00.0 vfio-pci

mkdir /mnt/huge
mount /mnt/huge --source nodev -t hugetlbfs -o pagesize=1GB
EOF
----

> . Customize the base OS image by using the `virt-customize` tool:

Reviewer 2: I'm not sure if "base OS image" is correct. I can't find any instances of it in the docs. Might it be `baseOSContainerImage`? 

MAYBE it's right. You might want to confirm with a dev. It definitely could be base OS image.

Author: Changed "base OS image" to "system image"

+
[source,terminal]
----
$ virt-customize -a <UUID>.qcow2 --run=customize-vm --firstboot=first-boot --selinux-relabel
----

. Create a Dockerfile that contains all the commands to build the container disk image:
+
[source,terminal]
----
$ cat << EOF > Dockerfile
FROM scratch
COPY <uuid>-disk.qcow2 /disk/
EOF
----

> . Build and tag the container:

Reviewer 2: ```suggestion
. Enter the following command to build and tag the container:
```

+
[source,terminal]
----
$ podman build . -t dpdk-rhel:latest
----

> . Push the container disk image to a registry that is accessible to your cluster:

Reviewer 2: ```suggestion
. Enter the following command to push the container disk image to a registry that is accessible to your cluster:
```

// Module included in the following assemblies:
//
// * virt//support/monitoring/virt-running-cluster-checkups.adoc              

:_content-type: PROCEDURE                                             
[id="virt-building-vm-containerdisk-image_{context}"]                                  
= Building a container disk image for {op-system-base} virtual machines

You can build a custom {op-system-base-full} 8 OS image in `qcow2` format and use it to create a container disk image. You can store the container disk image in a registry that is accessible from your cluster and specify the image location in the `spec.param.vmContainerDiskImage` attribute of the DPDK checkup config map. 

To build a container disk image, you must create an image builder virtual machine (VM). The _image builder VM_ is a {op-system-base} 8 VM that can be used to build custom {op-system-base} images.

.Prerequisites
* You have installed the image builder tool and its CLI (`composer-cli`).

> * The image builder VM must run {op-system-base} 8.7 and must have a minimum of 2 CPU cores, 4 GiB RAM, and 20 GB of free space in the `/var` directory.

Reviewer: Could you please switch the order of these sentences?
The image builder tool is install inside the VM.

----

. Generate the system image by specifying the blueprint name and output file format. The Universally Unique Identifier (UUID) of the image is displayed when you start the compose process.
+
[source,terminal]
----
# composer-cli compose start dpdk_image qcow2
----

. Wait for the compose process to complete. The compose status must show `FINISHED` before you can continue to the next step.
+
[source,terminal]
----
# composer-cli compose status
----

. Enter the following command to download the `qcow2` image file by specifying its UUID:
+
[source,terminal]
----
# composer-cli compose image <UUID>
----

. Create the customization scripts by running the following commands:
+
[source,terminal]
----
$ cat <<EOF >customize-vm
echo  isolated_cores=2-7 > /etc/tuned/cpu-partitioning-variables.conf
tuned-adm profile cpu-partitioning
echo "options vfio enable_unsafe_noiommu_mode=1" > /etc/modprobe.d/vfio-noiommu.conf
EOF
----
+
[source,terminal]
----
$ cat <<EOF >first-boot
driverctl set-override 0000:06:00.0 vfio-pci
driverctl set-override 0000:07:00.0 vfio-pci

mkdir /mnt/huge
mount /mnt/huge --source nodev -t hugetlbfs -o pagesize=1GB
EOF
----

> . Customize the system image by using the `virt-customize` tool:

Reviewer: ```suggestion
. Customize the image produced by `image builder` using the `virt-customize` tool:
```

+
[source,terminal]
----
$ virt-customize -a <UUID>.qcow2 --run=customize-vm --firstboot=first-boot --selinux-relabel
----

. To create a Dockerfile that contains all the commands to build the container disk image, enter the following command:
+
[source,terminal]
----
$ cat << EOF > Dockerfile
FROM scratch

> COPY <uuid>-disk.qcow2 /disk/

Reviewer: Please consider adding a note about editing this line to the proper `qcow2` image name.
